### ISSUE TITLE
Possibly missing ability to control DNS masquerading via Vagrantfile

### ISSUE THREAD

**Issue by jmfrank63** [vagrant-libvirt/vagrant-libvirt#1521]
**Is your feature request related to a problem? Please describe.**
In my company, I use vagrant to run our integration tests locally, which needs two clean Ubuntu servers to fully run.
The main software is running locally, and the two virtual machines are the target machines.
Since locally there might be a wireless interface on the local LAN, we cannot use a bridged network, but NAT works just fine.
However, NAT always listens on port 53 on the host. This conflicts with our software, which does try to install a DNS and thus fails because of the port conflict on the two NAT interfaces.
I managed to get it working by creating the management and our additional private network setup via `virsh` and then having vagrant use these networks. The networks use `<dns enable="no"/>.
While there are solutions alongside local DNS masquerading, there appears to be no obvious way to either disable DNS resolution, or move the port to a custom value.

**Describe the solution you'd like**
I would like to have the ability to control the DNS port and provide a custom value.
`:libvirt__dns_enabled` with defaults to true so not to break anything.
`libvirt__dns_port` which defaults to 53 for the same reason.

**Describe alternatives you've considered**
Creating the networks beforehand works, but is cumbersome. You also might have to consider disallowing destroying by vagrant.

**Additional context**
I create this issue for I, provided there isn't already a solution for this available, I consider adding the feature myself and contribute.
But beforehand I'd like to hear any opinions about this and possible drawbacks of the two proposed parameters, or take suggestions on the naming.
I am not a full ruby developer, but managed to contribute to the parallels plugin before, so with some guidance, I think I'll manage.

**Comment by electrofelix** [comment]
This sounds like you want to be able to disable DNS as I didn't see any useful way to modify the ports used via libvirt to run dnsmasq.

See https://libvirt.org/formatnetwork.html#addressing for details on how to configure a DNS block on a network definition which can be set to use `enable=no` to prevent libvirt from creating the dnsmasq process on that network.

I expect you'll want to add this to the private network template, worth adding a validation check if `domain_name` has been set as if DNS is disabled on the libvirt network, setting domain name should be considered an error in case there is an incorrect assumption that it should have an effect by the user.

**Comment by jmfrank63** [comment]
@electrofelix Thanks for the prompt answer. Do I understand this correctly, that it would be fine for me to work on a PR on that matter? I did consult the docs you mentioned above, in fact, that is where I got the info on how to disable it from the `virsh` side. However, I haven't found any way to achieve the same from the vagrant file.
Did I miss something?

**Comment by electrofelix** [comment]
@jmfrank63 yes, it seems that you've a good use case. No, there is currently no way to handle this from vagrant-libvirt.

If my understanding of the libvirt docs is correct, I believe it would require a `dns` element to be added to the private network template, somewhere around the sections:
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/c5ed4bd89c7a508825fda91ef3c86bd435b9be50/lib/vagrant-libvirt/templates/private_network.xml.erb#L13-L39

The tricky part with working on this is that there are still no tests in place for the networking parts of vagrant-libvirt. Since that would be pretty involved, I don't think it makes sense to try and solve as part of this unless you feel like you could add some. However it should be possible to add some checks for just the template itself without being too onerous by following the existing example of what has been done for the domain template.
https://github.com/vagrant-libvirt/vagrant-libvirt/tree/master/spec/unit/templates contains the existing tests for the domain template. 

It won't prevent mistakes in the networking actions but it will help make it easier to identify what needs to be adjusted on the template itself.

**Comment by electrofelix** [comment]
Actually it might make more sense to add a test for the behaviour you are looking to add to the acceptance tests under https://github.com/vagrant-libvirt/vagrant-libvirt/tree/master/spec/acceptance

Combination of https://github.com/vagrant-libvirt/vagrant-libvirt/blob/c5ed4bd89c7a508825fda91ef3c86bd435b9be50/spec/acceptance/use_qemu_agent_for_connectivity_spec.rb and https://github.com/vagrant-libvirt/vagrant-libvirt/blob/c5ed4bd89c7a508825fda91ef3c86bd435b9be50/spec/acceptance/support-skeletons/qemu_agent/Vagrantfile which contains a basic private network definition should provide a starting point to be able to add something that checks if the dns configuration is working as expected. Since that example shows how to disable the management network being attached it should make it easier to be sure that any dns request from within the VM isn't being satisfied by the mgmt network and is only using the custom config.

**Comment by johannes-enhance** [comment]
@electrofelix Sorry for the late reply. I have been struck by a real bad bug, thus the silence. I will take some reading and come back with my questions. Thanks for the support so far.

**Comment by jmfrank63** [comment]
@electrofelix Sorry for the late reply. I had been struck by a real bad bug. I will fork and create a branch, and after some reading trying to implement the tests you spoke of. Should I get stuck, then I'll come back ;-)
